src/factory.ts

```diff
@@ -0,0 +1,104 @@
+ * produced at a given rate. Recursively call this function to produce all ingredients.
+ * @param factory FactoryGraph to be modified
+ * @param item Item to be built
+ * @param rate Rate at which to build the product
+ */
+export function buildDependencies(
+    factory: FactoryGraph,
+    item: any,
+    rate: PerMinute,
+): ContainerNode {
+    /** Get containers holding this product */
+    const containers = factory.getContainers(item)
+    /** Increase egress of existing container if possible. No need to increase production of dependencies */
+    for (const container of containers) {
+        if (
+            container.getEgress() + rate <= container.getIngress() &&
+            container.outgoingLinkCount < 10
+        ) {
+            return container
+        }
+    }
+    let output: ContainerNode | undefined = undefined
+    /** Add more producers if possible */
+    const recipe = findRecipe(item)
+    for (const container of containers) {
+        /** Required number of additional producers */
+        const count = Math.ceil(
+            (rate + (container.getEgress() - container.getIngress())) /
+                (recipe.product.quantity / recipe.time),
+        )
+        if (container.incomingLinkCount + count <= 10 && container.outgoingLinkCount < 10) {
+            output = container
+        }
+    }
+    /** Create a new container if necessary */
+    if (output === undefined) {
+        output = new ContainerNode(item)
+        factory.addContainer(output)
+    }
+    /** Add new producers */
+    const count = Math.ceil(
+        (rate + (output.getEgress() - output.getIngress())) /
+            (recipe.product.quantity / recipe.time),
+    )
+    for (let i = 0; i < count; i++) {
+        const industry = new IndustryNode(item)
+        industry.outputTo(output)
+        factory.addIndustry(industry)
+        /** Build dependencies recursively */
+        for (const ingredient of recipe.ingredients) {
+            /** If ingredient is an ore, link from new or existing ore container */
+            if (isOre(ingredient.item)) {
+                const oreContainers = Array.from(factory.getContainers(ingredient.item))
+                let input: ContainerNode | undefined = undefined
+                if (oreContainers.length > 0) {
+                    /* Use existing ore container */
+                    input = oreContainers[0]
+                } else {
+                    /* Create new ore container */
+                    input = new ContainerNode(ingredient.item)
+                    factory.addContainer(input)
+                }
+                industry.takeFrom(input, ingredient.item)
+            } else {
+                const input = buildDependencies(
+                    factory,
+                    ingredient.item,
+                    ingredient.quantity / recipe.time,
+                )
+                industry.takeFrom(input, ingredient.item)
+            }
```

Comment:
We could move the `if` block to the start of the function and only leave the `else` block here. That way we don't need an extra check above if the item is an ore.